crates/meilitool/src/upgrade/v1_12.rs

```diff
@@ -1,24 +1,58 @@
 //! The breaking changes that happened between the v1.11 and the v1.12 are:
 //! - The new indexer changed the update files format from OBKV to ndjson. https://github.com/meilisearch/meilisearch/pull/4900
 
+use std::borrow::Cow;
 use std::io::BufWriter;
 use std::path::Path;
+use std::sync::atomic::AtomicBool;
 
 use anyhow::Context;
 use file_store::FileStore;
 use indexmap::IndexMap;
 use meilisearch_types::milli::documents::DocumentsBatchReader;
+use milli::heed::types::{SerdeJson, Str};
+use milli::heed::{Database, EnvOpenOptions, RoTxn, RwTxn};
+use milli::progress::Step;
+use milli::{FieldDistribution, Index};
+use serde::Serialize;
 use serde_json::value::RawValue;
 use tempfile::NamedTempFile;
+use time::OffsetDateTime;
+use uuid::Uuid;
 
-pub fn v1_11_to_v1_12(db_path: &Path) -> anyhow::Result<()> {
+use crate::try_opening_database;
+use crate::uuid_codec::UuidCodec;
+
+pub fn v1_11_to_v1_12(
+    db_path: &Path,
+    _origin_major: &str,
+    _origin_minor: &str,
+    _origin_patch: &str,
+) -> anyhow::Result<()> {
     println!("Upgrading from v1.11.0 to v1.12.0");
 
     convert_update_files(db_path)?;
 
     Ok(())
 }
 
+pub fn v1_12_to_v1_12_3(
+    db_path: &Path,
+    origin_major: &str,
+    origin_minor: &str,
+    origin_patch: &str,
+) -> anyhow::Result<()> {
+    println!("Upgrading from v1.12.{{0, 1, 2}} to v1.12.3");
+
+    if origin_minor == "12" {
```

Comment:
If this function is called, then we're guaranteed that the previous version was the v1.12.{0,1,2}.
We don't need this check, and we don't need to retrieve the previous version, I think 🤔

Reply:
There are two cases here:

1. This function is called in a chain of calls coming from e.g. v1.10 to v1.12.3:
    1. v1.10 to v1.11
    2. v1.11 to v1.12
    3. v1.12 to v1.12.3
2. This function is directly called from v1.12.x:
    1. v1.12 to v1.12.3

We only want to rebuild facet distribution in case (2), as it wasn't corrupted in case (1).
The functions were updated to accept the `origin_major`, `origin_minor`, `origin_patch`, which are the version of the DB **before any upgrade operation started**.

Thus, it should allow us to distinguish between case (1) and case (2)

Reply:
ah yes got it! 👍